File: Python/Merge Intervals.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    """
    :type intervals: List[Interval]
    :rtype: List[Interval]
    """
    def merge(self, intervals):
        if intervals == None or len(intervals) <= 1:
            return intervals
    
        intervals = sorted(intervals, key=lambda interval : interval.start)
        
        result = []
        interval = intervals[0]
        for i in range(1, len(intervals), 1):
            cur = intervals[i]
            if cur.start <= interval.end:
                interval.end = max(interval.end, cur.end)
            else:
                result.append(interval)
                interval = cur
                        
                        
        result.append(interval)
        return result

File: Java/Merge Intervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> result = new ArrayList<Interval>();
        if (intervals == null || intervals.size() == 0)
            return result;
        
        Collections.sort(intervals, new sortByStart());
        
        Interval i = intervals.get(0);
        for (int j=1; j<intervals.size(); j++) {
            Interval next = intervals.get(j);
            if (next.start<=i.end) {
                i.end = Math.max(next.end, i.end);
            } else {
                result.add(i);
                i = next;
            }
        }
        result.add(i);
        return result;
    }
    
    class sortByStart implements Comparator<Interval> {
        public int compare(Interval i1, Interval i2) {
            return i1.start - i2.start;
        }
    }
}